### PR TITLE
redolog: add a precleanup process when s3 enable

### LINKED ISSUE #3523: redo log writer runGC failed because of `can't read log file directory`
### What did you do?

1. Create a tidb cluster, start up a TiCDC
2. create a changefeed by `cdc cli changefeed create --pd=http://127.0.0.1:2379 -c bypass-cf-5 --sink-uri="blackhole://" --config changefeed5_nfs.toml`
3. changefeed5_nfs.toml

```
[consistent]
level = "eventual"
storage = "nfs:///home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5"
```

Path `/home/apple/work/code/pingcap/ticdc/bin/redo/` exists, but `/home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5` doesn't.

### What did you expect to see?

cdc runs normally

### What did you see instead?

cdc keeps logging

```
[2021/11/18 17:17:52.891 +08:00] [ERROR] [writer.go:237] ["redo log GC fail"] [changefeedID=bypass-cf-5] [error="[CDC:ErrRedoFileOp]can't read log file directory: /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: open /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: no such file or directory: can't read log file directory: /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: open /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: no such file or directory; [CDC:ErrRedoFileOp]can't read log file directory: /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: open /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: no such file or directory: can't read log file directory: /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: open /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: no such file or directory"] [errorVerbose="the following errors occurred:\n -  [CDC:ErrRedoFileOp]can't read log file directory: /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: open /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: no such file or directory: can't read log file directory: /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: open /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: no such file or directory\n    github.com/pingcap/errors.AddStack\n    \tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/errors.go:174\n    github.com/pingcap/errors.(*Error).GenWithStackByCause\n    \tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/normalize.go:302\n    github.com/pingcap/ticdc/pkg/errors.WrapError\n    \tgithub.com/pingcap/ticdc/pkg/errors/helper.go:30\n    github.com/pingcap/ticdc/cdc/redo/writer.(*Writer).getShouldRemovedFiles\n    \tgithub.com/pingcap/ticdc/cdc/redo/writer/file.go:462\n    github.com/pingcap/ticdc/cdc/redo/writer.(*Writer).GC\n    \tgithub.com/pingcap/ticdc/cdc/redo/writer/file.go:411\n    github.com/pingcap/ticdc/cdc/redo/writer.(*LogWriter).gc\n    \tgithub.com/pingcap/ticdc/cdc/redo/writer/writer.go:249\n    github.com/pingcap/ticdc/cdc/redo/writer.(*LogWriter).runGC\n    \tgithub.com/pingcap/ticdc/cdc/redo/writer/writer.go:235\n    runtime.goexit\n    \truntime/asm_amd64.s:1371\n -  [CDC:ErrRedoFileOp]can't read log file directory: /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: open /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: no such file or directory: can't read log file directory: /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: open /home/apple/work/code/pingcap/ticdc/bin/redo/bypass-cf-5: no such file or directory\n    github.com/pingcap/errors.AddStack\n    \tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/errors.go:174\n    github.com/pingcap/errors.(*Error).GenWithStackByCause\n    \tgithub.com/pingcap/errors@v0.11.5-0.20211009033009-93128226aaa3/normalize.go:302\n    github.com/pingcap/ticdc/pkg/errors.WrapError\n    \tgithub.com/pingcap/ticdc/pkg/errors/helper.go:30\n    github.com/pingcap/ticdc/cdc/redo/writer.(*Writer).getShouldRemovedFiles\n    \tgithub.com/pingcap/ticdc/cdc/redo/writer/file.go:462\n    github.com/pingcap/ticdc/cdc/redo/writer.(*Writer).GC\n    \tgithub.com/pingcap/ticdc/cdc/redo/writer/file.go:411\n    github.com/pingcap/ticdc/cdc/redo/writer.(*LogWriter).gc\n    \tgithub.com/pingcap/ticdc/cdc/redo/writer/writer.go:250\n    github.com/pingcap/ticdc/cdc/redo/writer.(*LogWriter).runGC\n    \tgithub.com/pingcap/ticdc/cdc/redo/writer/writer.go:235\n    runtime.goexit\n   
 \truntime/asm_amd64.s:1371"]
```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
v5.3.0
```

TiCDC version (execute `cdc version`):

master@https://github.com/pingcap/ticdc/commit/fe92b89a0ea05a066a61a94d96440f38504d170e

### PR DESCRIPTION
**Pull request by ben1009**
### What problem does this PR solve? 

- add a precleanup prcess when s3 enable
- close https://github.com/pingcap/ticdc/issues/3523

### What is changed and how it works?

- set a delete marker when delete logs in s3, since the  process got the remove changefeed job async, may continue write logs after owner remove the logs. if have the changefeed with the same name created later, pre cleanup the log just in case.

- when gc if the dir is not exist, just log instead of return err

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

Related changes

### Release note 

```release-note
add pre clean up process when s3 enable, fix #3523
```

### CODE REVIEW

File: cdc/redo/writer/writer.go [hunk 1]
```diff
@@ -433,7 +473,18 @@ func (l *LogWriter) DeleteAllLogs(ctx context.Context) error {
...
+	return l.writeDeletedMarkerToS3(ctx)
+}
+
+func (l *LogWriter) getDeletedChangefeedMarker() string {
+	return fmt.Sprintf("delete_%s", l.cfg.ChangeFeedID)
+}
+
+func (l *LogWriter) writeDeletedMarkerToS3(ctx context.Context) error {
+	return cerror.WrapError(cerror.ErrS3StorageAPI, l.storage.WriteFile(ctx, l.getDeletedChangefeedMarker(), []byte("D")))
 }
```
Comment (amyangfei) [inline cdc/redo/writer/writer.go:488]: Is the delete marker file written to S3, and the redo log dir in S3 is not deleted after changefeed is removed?

Comment (ben1009) [reply cdc/redo/writer/writer.go:488]: yes, the bucket is still there, no extra cost. but currently we do not delete the bucket even without the marker, just delete the object

File: cdc/redo/writer/writer_test.go [hunk 2]
```diff
@@ -839,3 +876,73 @@ func TestDeleteAllLogs(t *testing.T) {
...
+			wantErr:       ".*ErrS3StorageAPI*.",
+		},
+		{
+			name:          "deleteFile notExist err",
+			fileExists:    true,
+			deleteFileErr: awserr.New(s3.ErrCodeNoSuchKey, "no such key", nil),
+		},
+	}
+
+	for _, tt := range tests {
```
Comment (hi-rustin) [inline cdc/redo/writer/writer_test.go:917]:
```suggestion
	for _, tc := range testCases {
```

File: cdc/redo/writer/writer_test.go [hunk 3]
```diff
@@ -839,3 +876,73 @@ func TestDeleteAllLogs(t *testing.T) {
...
+		origin := getAllFilesInS3
+		getAllFilesInS3 = func(ctx context.Context, l *LogWriter) ([]string, error) {
+			return []string{"1", "11", "delete_test-cf"}, tt.getAllFilesInS3Err
+		}
+		controller := gomock.NewController(t)
+		mockStorage := mockstorage.NewMockExternalStorage(controller)
+
+		mockStorage.EXPECT().FileExists(gomock.Any(), gomock.Any()).Return(tt.fileExists, tt.fileExistsErr)
```
Comment (hi-rustin) [inline cdc/redo/writer/writer_test.go:925]: Can we just match the error without mock?

Comment (ben1009) [reply cdc/redo/writer/writer_test.go:925]: `mockStorage.EXPECT()` have a EXPECT  the func been called beside the return value

Comment (hi-rustin) [reply cdc/redo/writer/writer_test.go:925]: Got it.

**Review by hi-rustin — APPROVED** [review]

**Review by amyangfei — APPROVED** [review]

**Comment by ben1009** [comment]
> /run-kafka-integration-test /run-dm-integration-test

+1

**Comment by ben1009** [comment]
> /run-integration-test
> /run-dm-integration-test
> 
> 
>

**Comment by ben1009** [comment]
> /run-verify

1

**Comment by ben1009** [comment]
> /run-integration-test

1

**Pull request merged**